File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Deque<Integer> stack = new ArrayDeque<>();

        for (int asteroid : asteroids) {
            boolean isAlive = true;

            while (!stack.isEmpty() && stack.peek() > 0 && asteroid < 0) {
                int top = stack.peek();
                if (top < -asteroid) {
                    stack.pop();
                } else if (top == -asteroid) {
                    stack.pop();
                    isAlive = false;
                    break;
                } else {
                    isAlive = false;
                    break;
                }
            }

            if (isAlive) stack.push(asteroid);
        }

        int[] result = new int[stack.size()];
        for (int i = stack.size() - 1; i >= 0; i--) {
            result[i] = stack.pop();
        }

        return result;
    }
}

File: 0735-asteroid-collision/0735-asteroid-collision.kt
class Solution {
    fun asteroidCollision(asteroids: IntArray): IntArray {
        val stack = ArrayDeque<Int>()

        for (asteroid in asteroids) {
            var isAlive = true

            while (stack.isNotEmpty() && stack.last() > 0 && asteroid < 0) {
                val top = stack.last()
                when {
                    top < -asteroid -> stack.removeLast()
                    top == -asteroid -> {
                        stack.removeLast()
                        isAlive = false
                        break
                    }
                    else -> {
                        isAlive = false
                        break
                    }
                }
            }

            if (isAlive) stack.addLast(asteroid)
        }

        return stack.toIntArray()
    }
}